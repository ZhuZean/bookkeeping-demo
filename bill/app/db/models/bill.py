""" Declarative model for Bill
"""

from enum import Enum
import datetime
import uuid

import sqlalchemy as sa
from sqlalchemy import orm
from sqlalchemy.dialects import postgresql
from sqlalchemy.orm import relationship

from app.db.models import base_model
from app.db.models.payment import Payment
from app.db.models.usage import Usage
from app.db.models.currency import Currency


class BillTypeEnum(str, Enum):
    income = 'income'
    expense = 'expense'


class Bill(base_model.BASE):
    """ Model Bill
    """
    __tablename__ = 'bill'
    id = sa.Column(
        postgresql.UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        nullable=False
    )

    bill_type = sa.Column(sa.String, nullable=False)
    note = sa.Column(sa.Text, nullable=True)
    price = sa.Column(sa.Numeric(asdecimal=True), nullable=False)

    payment_id = sa.Column(postgresql.UUID(as_uuid=True), sa.ForeignKey('payment.id'), nullable=False, index=True)
    payment = relationship('Payment')

    usage_id = sa.Column(postgresql.UUID(as_uuid=True), sa.ForeignKey('usage.id'), nullable=False, index=True)
    usage = relationship('Usage')

    currency_id = sa.Column(postgresql.UUID(as_uuid=True), sa.ForeignKey('currency.id'), nullable=False, index=True)
    currency = relationship('Currency')

    created_at = sa.Column(
        sa.DateTime(timezone=True), nullable=True, default=datetime.datetime.utcnow
    )
    updated_at = sa.Column(
        sa.DateTime(timezone=True), nullable=True, default=datetime.datetime.utcnow, onupdate=datetime.datetime.utcnow
    )
