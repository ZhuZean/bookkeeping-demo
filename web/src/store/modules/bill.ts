import billService from '@/service/bill-service'
import { getField, updateField } from 'vuex-map-fields'

type BillDetail = {
    'payment': any,
    'bill_type': string,
    'usage': any,
    'currency': any,
    'price': number,
    'note': string,
    'updated_at': string,
}

type Pagination = {
    'count': number,
    'next': string | null,
    'previous': string | null,
    'total_pages': number
}

type BillSummary = {
    'total_expense': number,
    'total_income': number
}

type Payment = {
    'name': string,
    'amount': number
}

type BillInfo = {
    'payment': any[],
    'usage': any[],
    'currency': any[],
    'bill_type': string[]
}

type BillForm = {
    'payment': string,
    'billType': string,
    'usage': string,
    'currency': string,
    'price': number,
    'note': string
}

function isEmptyObject(obj: any) {
    return Object.keys(obj).length === 0
}

function getBillFormRequestData(billForm: BillForm) {
    console.log(billForm)
    return {
        "payment_id": billForm.payment,
        "bill_type": billForm.billType,
        "usage_id": billForm.usage,
        "currency_id": billForm.currency,
        "price": billForm.price,
        "note": billForm.note
      }
}

const getters = {
    bills: (state: any) => {
        return state.bills
    },
    pagination: (state: any) => {
        return state.pagination
    },
    summary: (state: any) => {
        return state.summary
    },
    paymentDetail: (state: any) => {
        return state.paymentDetail
    },
    billInfo: (state: any) => {
        return state.billInfo
    },
    getAddBillFormField(state: any) {
        return getField(state.addBillForm)
    },
    addBillFormErrors: (state: any) => {
        return state.addBillFormErrors
    },
    billId:  (state: any) => {
        return state.billId
    }
}

const actions = {
    initState(context: any) {
        Object.assign(context.state, initialState())
    },
    async getBills(context: any, payload: any) {
        const resp = await billService.getBills(payload)
        context.commit('setBills', resp.items)
        context.commit('setPagination', resp.pagination)
    },
    async getBillSummary(context: any) {
        const resp = await billService.getBillSummary()
        context.commit('setSummary', resp.summary)
        context.commit('setPaymentDetail', resp.payment_detail)
    },
    async getBillInfo(context: any) {
        const resp = await billService.getBillInfo()
        context.commit('setBillInfo', resp)
    },
    async addBill(context: any) {
        // do validation
        context.dispatch('bill/_validate_form', null, { root: true })

        // terminate if any form error exists
        const commonFormErrors = context.getters.addBillFormErrors
        if (!isEmptyObject(commonFormErrors)) {
            throw new Error('Add bill form is invalid')
        }

        const billFormData = context.state.addBillForm
        const resp = await billService.addBill(getBillFormRequestData(billFormData))
        context.commit('setBillId', resp.id)
    },
    _validate_form(context: any) {
        const errors: any = {}

        if (!context.state.addBillForm.payment.length) {
            errors.payment = true
        }

        if (!context.state.addBillForm.billType.length) {
            errors.billType = true
        }

        if (!context.state.addBillForm.usage.length) {
            errors.usage = true
        }

        if (!context.state.addBillForm.currency.length) {
            errors.currency = true
        }

        if (!context.state.addBillForm.price.toString().length || parseInt(context.state.addBillForm.price) < 0) {
            errors.price = true
        }

        context.commit('setFormErrors', errors)
    }
}

const mutations = {
    setBills(state: any, bills: BillDetail[]) {
        state.bills = bills
    },
    setPagination(state: any, pagination: Pagination) {
        state.pagination = pagination
    },
    setSummary(state: any, summary: BillSummary) {
        state.summary = summary
    },
    setPaymentDetail(state: any, paymentDetail: Payment[]) {
        state.paymentDetail = paymentDetail
    },
    setBillInfo(state: any, billInfo: BillInfo) {
        // set billInfo
        state.billInfo.payment = billInfo.payment
        state.billInfo.billType = billInfo.bill_type
        state.billInfo.usage = billInfo.usage
        state.billInfo.currency = billInfo.currency
        // set bill form default value
        state.addBillForm.payment = billInfo.payment[0].id
        state.addBillForm.billType = billInfo.bill_type[1]
        state.addBillForm.usage = billInfo.usage[0].id
        state.addBillForm.currency = billInfo.currency[0].id
    },
    updateAddBillFormField(state: any, field: any) {
        updateField(state.addBillForm, field);
    },
    setFormErrors(state: any, addBillFormErrors: any) {
        state.addBillFormErrors = addBillFormErrors
    },
    setBillId(state: any, billId: number) {
        state.billId = billId
    },
}

function initialState() {
    return {
        billId: 0,
        bills: [],
        pagination: {
            total_number_of_items: 0,
            current_page: 1,
            next_page: null,
            previous_page: null,
            total_pages: 0
        },
        summary: {
            total_expense: 0,
            total_income: 0
        },
        paymentDetail: [],
        billInfo: {
            payment: [],
            usage: [],
            currency: [],
            billType: []
        },
        addBillForm: {
            payment: '',
            billType: '',
            usage: '',
            currency: '',
            price: 0,
            note: ''
        },
        addBillFormErrors: {}
    }
}

export default {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: initialState()
}
