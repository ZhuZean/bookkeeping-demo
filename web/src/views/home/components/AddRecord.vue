<template>
  <div id="record" class="container">
    <form>
      <div class="row">
        <div class="col">
          <select class="form-control" v-model="payment">
            <option
              v-for="(paymentName, index) in billInfo.payment"
              :key="index"
              v-bind:value="paymentName"
            >
              {{ paymentName }}
            </option>
          </select>
        </div>
        <div class="col">
          <select class="form-control" v-model="billType">
            <option
              v-for="(billTypeName, index) in billInfo.billType"
              :key="index"
              v-bind:value="billTypeName"
            >
              {{ billTypeName }}
            </option>
          </select>
        </div>
        <div class="col">
          <select class="form-control" v-model="usage">
            <option
              v-for="(usageName, index) in billInfo.usage"
              :key="index"
              v-bind:value="usageName"
            >
              {{ usageName }}
            </option>
          </select>
        </div>
      </div>
      <div class="row mt-3">
        <div class="col-4">
          <select class="form-control" v-model="currency">
            <option
              v-for="(currencyName, index) in billInfo.currency"
              :key="index"
              v-bind:value="currencyName"
            >
              {{ currencyName }}
            </option>
          </select>
        </div>
        <div class="col">
          <input
            type="number"
            min="0"
            class="form-control"
            placeholder=""
            v-model="price"
            v-bind:class="{ 'is-invalid': addBillFormErrors.price }"
          />
          <div class="invalid-feedback">{{ $t("Home.AddRecord.form.price.error") }}</div>
        </div>
      </div>
      <div class="form row mt-3">
        <div class="col">
          <label for="notes"> {{ $t("Home.AddRecord.form.notes.label") }}</label>
          <input
            type="text"
            class="form-control"
            id="notes"
            placeholder="note"
            v-model="note"
          />
        </div>
      </div>
      <div class="text-right mt-4">
        <button type="button" class="btn btn-primary" @click="addBill()">
          {{ $t("Home.AddRecord.form.submit_btn") }}
        </button>
      </div>
    </form>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions } from "vuex";
import { Options, Vue } from "vue-class-component";
import { createHelpers } from "vuex-map-fields";

const { mapFields } = createHelpers({
  getterType: "bill/getAddBillFormField",
  mutationType: "bill/updateAddBillFormField",
});

@Options({
  name: "AddRecord",
  data() {
    return {};
  },
  computed: {
    ...mapGetters({
      billInfo: "bill/billInfo",
      addBillForm: "bill/addBillForm",
      addBillFormErrors: "bill/addBillFormErrors",
      billId: 'bill/billId'
    }),
    ...mapFields({
      payment: "payment",
      billType: "billType",
      usage: "usage",
      currency: "currency",
      price: "price",
      note: "note",
    }),
  },
  watch: {
    billId: function() {
      this.price = 0
      this.note = ''
    }
  },
  methods: {
    ...mapActions({
      getBillInfo: "bill/getBillInfo",
      addBill: "bill/addBill",
    }),
  },
  mounted: async function mounted() {
    await this.getBillInfo();
  },
})
export default class Home extends Vue {}
</script>
