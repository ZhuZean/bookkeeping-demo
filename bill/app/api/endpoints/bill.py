"""Endpoint for bill
"""
from typing import Any, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Body, HTTPException, Query

from app.api import deps
from app.api.endpoints import bad_request, business_error
from app.helper.paginator import Paginator
from app.schemas.bill import (
    CreateBillResponse,
    CreateBillRequest,
    ListBillResponse,
    BillFormInfoResponse,
    BillSummaryResponse
)


router = APIRouter()


@router.post("", response_model=CreateBillResponse, status_code=201)
def create(
    params: CreateBillRequest = Body(...),
    db=Depends(deps.get_db_session),
    bill_business=Depends(deps.get_bill_business)
) -> Any:
    """ Create bill
    """
    try:
        return bill_business.create(params=params, db=db)
    except ValueError as ex:
        raise bad_request({'error': str(ex)})
    except Exception as ex:
        raise business_error({'error': str(ex)})


@router.get("", response_model=ListBillResponse)
def list_bills(
    page: int = Query(1),
    limit: int = Query(5),
    db=Depends(deps.get_db_session),
    bill_business=Depends(deps.get_bill_business)
) -> Any:
    """ Create bill
    """
    try:
        paginator = Paginator(limit, page)
        items = bill_business.get_bills(paginator=paginator, db=db)
        return {'items': items, 'pagination': paginator.get_info()}
    except Exception as ex:
        raise business_error({'error': str(ex)})


@router.get("/form-info", response_model=BillFormInfoResponse)
def form_info(
    db=Depends(deps.get_db_session),
    bill_business=Depends(deps.get_bill_business)
) -> Any:
    """ Get form field information
    """
    try:
        return bill_business.get_form_info(db=db)
    except Exception as ex:
        raise business_error({'error': str(ex)})


@router.get("/summary", response_model=BillSummaryResponse)
def summary(
    db=Depends(deps.get_db_session),
    bill_business=Depends(deps.get_bill_business)
) -> Any:
    """ Get bill summary data
    """
    try:
        return bill_business.get_summary(db=db)
    except Exception as ex:
        raise business_error({'error': str(ex)})