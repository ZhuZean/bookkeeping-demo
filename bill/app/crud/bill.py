from typing import List

from sqlalchemy.orm import Session

from app.crud.base import CRUDBase
from app.db.models.bill import Bill as BillOrm
from app.schemas.orm.bill import BillCreate, BillUpdate


class CRUDBill(CRUDBase[BillOrm, BillCreate, BillUpdate]):
    def get_list(self, db: Session) -> List[BillOrm]:
        return db.query(self.model).order_by(self.model.updated_at.desc())


bill = CRUDBill(BillOrm)
