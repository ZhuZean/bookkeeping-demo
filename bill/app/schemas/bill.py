from typing import List
from decimal import Decimal

from pydantic import BaseModel

from app.schemas.orm.bill import BaseBill
from app.schemas.orm.payment import BasePayment
from app.schemas.orm.usage import BaseUsage
from app.schemas.orm.currency import BaseCurrency
from .pagination import PaginationResponse


class CreateBillRequest(BaseBill):
    pass


class CreateBillResponse(BaseBill):
    bill_type: str
    note: str
    price: Decimal
    payment: BasePayment
    usage: BaseUsage
    currency: BaseCurrency


class ListBillResponse(PaginationResponse):
    items: List[CreateBillResponse]

    class Config:
        orm_mode = True


class BillFormInfoResponse(BaseModel):
    payment: List[BasePayment]
    usage: List[BaseUsage]
    currency: List[BaseCurrency]
    bill_type: List[str]

    class Config:
        orm_mode = True


class BaseSummary(BaseModel):
    total_expense: Decimal
    total_income: Decimal


class BasePaymentDetail(BaseModel):
    name: str
    amount: Decimal


class BillSummaryResponse(BaseModel):
    summary: BaseSummary
    payment_detail: List[BasePaymentDetail]

    class Config:
        orm_mode = True
