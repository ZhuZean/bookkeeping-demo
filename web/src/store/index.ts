import { createStore } from 'vuex'
import bill from '@/store/modules/bill'

export default createStore({
  modules: {
    bill
  }
})
