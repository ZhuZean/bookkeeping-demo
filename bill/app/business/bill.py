""" Bill business
"""
from uuid import UUID

from sqlalchemy.orm import Session

from app.crud.bill import bill as crud_bill
from app.db.models.bill import Bill as BillOrm, BillTypeEnum
from app.db.models.currency import Currency as CurrencyOrm
from app.db.models.payment import Payment as PaymentOrm
from app.db.models.usage import Usage as UsageOrm
from app.schemas.bill import CreateBillRequest
from app.helper.paginator import Paginator


class BillBusiness():

    def create(self, params: CreateBillRequest, db: Session=None):
        bill = crud_bill.create(db=db, obj_in=params)

        return bill

    def get_bills(self, paginator: Paginator, db: Session=None):
        bill_query = crud_bill.get_list(db=db)

        return paginator.get_page(bill_query)

    def get_form_info(self, db: Session=None):
        bill_query = crud_bill.get_list(db=db)
        result = {
            'currency': self._get_all_currency_name(db),
            'payment': self._get_all_payment_name(db),
            'usage': self._get_all_usage_name(db),
            'bill_type': self._get_all_bill_type(),
        }

        return result

    def _get_all_currency_name(self, db: Session) -> list:
        currency_query = db.query(CurrencyOrm)
        all_currency = currency_query.order_by(CurrencyOrm.updated_at).all()

        return [currency.name for currency in all_currency]

    def _get_all_payment_name(self, db: Session) -> list:
        payment_query = db.query(PaymentOrm)
        all_payment = payment_query.order_by(PaymentOrm.updated_at).all()

        return [payment.name for payment in all_payment]

    def _get_all_usage_name(self, db: Session) -> list:
        usage_query = db.query(UsageOrm)
        all_usage = usage_query.order_by(UsageOrm.updated_at).all()

        return [usage.name for usage in all_usage]

    def _get_all_bill_type(self) -> list:
        return [bill_type.value for bill_type in BillTypeEnum]

    def get_summary(self, db: Session=None):
        payment_detail = self._get_payment_details(db)
        result = {
            'summary': self._get_summary(db),
            'payment_detail': self._get_payment_details(db)
        }
        return result

    def _get_summary(self, db: Session=None) -> dict:
        result = {
            'total_expense': 0,
            'total_income': 0
        }
        bill_queryset = db.query(
            BillOrm
        ).order_by(
            'created_at'
        ).all()

        for bill_obj in bill_queryset:
            if (bill_obj.bill_type == BillTypeEnum.expense):
                result['total_expense'] += bill_obj.price
            if (bill_obj.bill_type == BillTypeEnum.income):
                result['total_income'] += bill_obj.price
        return result

    def _get_payment_details(self, db) -> list:
        payment_queryset = db.query(
            PaymentOrm
        ).order_by(
            'created_at'
        ).all()

        payment_detail = []
        for payment_obj in payment_queryset:
            detail = {
                'name': payment_obj.name,
                'amount': self._get_payment_summary(db, payment_obj)
            }
            payment_detail.append(detail)

        return payment_detail

    def _get_payment_summary(self, db, payment) -> int:
        amount = 0

        bill_queryset = db.query(
            BillOrm
        ).filter(
            BillOrm.payment_id==payment.id
        ).order_by(
            'created_at'
        ).all()

        for bill_obj in bill_queryset:
            if (bill_obj.bill_type == BillTypeEnum.expense):
                amount -= bill_obj.price
            if (bill_obj.bill_type == BillTypeEnum.income):
                amount += bill_obj.price

        return amount
