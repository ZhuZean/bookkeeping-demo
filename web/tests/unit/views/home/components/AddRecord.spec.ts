// @ts-nocheck
import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'

import Vuex from 'vuex'

import localization from '../../../mocks/localization.mock'
import AddRecord from '@/views/home/components/AddRecord.vue'
import bill from '@/store/modules/bill'
import store from '@/store'


describe('Home.vue', () => {
  let store: any
  let components: any

  beforeEach(() => {
    store = createStore({
      modules: {
        bill
      }
    })

    components = {
      plugins: [store],
      mocks: {
        ...localization
      },
    }
  })

  it('can render', () => {
    const wrapper = shallowMount(AddRecord, {
      global: components
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('can watch billId', async () => {
    const wrapper = shallowMount(AddRecord, {
      global: components
    })
    // set the value of price and note
    wrapper.vm.price = 1000
    wrapper.vm.note = 'test'

    // commit to change the value of billId
    store.commit('bill/setBillId', 'new-id')
    await wrapper.vm.$nextTick()

    // watch the change of BillId and reset the value of billId
    expect(wrapper.vm.price).toBe(0)
    expect(wrapper.vm.note).toBe('')
    expect(wrapper.element).toMatchSnapshot()
  })

})
