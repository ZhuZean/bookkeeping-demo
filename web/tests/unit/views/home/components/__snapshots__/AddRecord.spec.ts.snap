// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue can render 1`] = `
<div
  class="container"
  id="record"
>
  <form>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <select
          class="form-control"
        >
          
          
        </select>
      </div>
      <div
        class="col"
      >
        <select
          class="form-control"
        >
          
          
        </select>
      </div>
      <div
        class="col"
      >
        <select
          class="form-control"
        >
          
          
        </select>
      </div>
    </div>
    <div
      class="row mt-3"
    >
      <div
        class="col-4"
      >
        <select
          class="form-control"
        >
          
          
        </select>
      </div>
      <div
        class="col"
      >
        <input
          class="form-control"
          min="0"
          placeholder=""
          type="number"
        />
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
    <div
      class="form row mt-3"
    >
      <div
        class="col"
      >
        <label
          for="notes"
        />
        <input
          class="form-control"
          id="notes"
          placeholder="note"
          type="text"
        />
      </div>
    </div>
    <div
      class="text-right mt-4"
    >
      <button
        class="btn btn-primary"
        type="button"
      />
    </div>
  </form>
</div>
`;

exports[`Home.vue can watch billId 1`] = `
<div
  class="container"
  id="record"
>
  <form>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <select
          class="form-control"
        >
          
          
        </select>
      </div>
      <div
        class="col"
      >
        <select
          class="form-control"
        >
          
          
        </select>
      </div>
      <div
        class="col"
      >
        <select
          class="form-control"
        >
          
          
        </select>
      </div>
    </div>
    <div
      class="row mt-3"
    >
      <div
        class="col-4"
      >
        <select
          class="form-control"
        >
          
          
        </select>
      </div>
      <div
        class="col"
      >
        <input
          class="form-control"
          min="0"
          placeholder=""
          type="number"
        />
        <div
          class="invalid-feedback"
        />
      </div>
    </div>
    <div
      class="form row mt-3"
    >
      <div
        class="col"
      >
        <label
          for="notes"
        />
        <input
          class="form-control"
          id="notes"
          placeholder="note"
          type="text"
        />
      </div>
    </div>
    <div
      class="text-right mt-4"
    >
      <button
        class="btn btn-primary"
        type="button"
      />
    </div>
  </form>
</div>
`;
